File: 2020/day08/python/day81.py
ops = [line.strip() for line in open("../input.txt")]
ops = [[False, line[:3], line[3:]] for line in ops]
#print(ops)

pc = 0
acc = 0
while(True):
    curr_op = ops[pc]
    
    visited, op_name, val = curr_op
    if visited : break
    ops[pc][0] = True
    pc += 1
    if op_name == "nop": continue
    if op_name == "acc": acc += int(val)
    if op_name == "jmp": 
        pc += int(val)-1
        continue
    
print(acc)

File: 2020/day08/haskell/day81.hs
import Data.Sequence (fromList, update, index, Seq)
main = interact $ show 
  . runProgram 0 0
  . fromList . zip (repeat False) . map parse . lines

runProgram :: Int -> Int -> Seq (Bool, (String, Int)) -> Int
runProgram acc n list
  | visited = acc
  | otherwise =
    case op_name of
      "nop" -> runProgram acc (n+1) newList
      "acc" -> runProgram (acc+num) (n+1) newList
      "jmp" -> runProgram acc (n+num) newList
  where
    (visited,(op_name, num)) = list `index` n
    newList = update n (True, (op_name, num)) list

parse :: String -> (String, Int)
parse = (\(op:snum:xs) -> (op, parseNum snum)) . words

parseNum :: String -> Int
parseNum ('+':num) = read num
parseNum ('-':num) = negate $ read num